specs/overview.md

```diff
@@ -0,0 +1,158 @@
+
+Unsubmitted blocks, however, are propagated using a separate peer-to-peer network of Rollup Nodes. This is optional, however, and is provided as a convnience to lower latency for users.
+
+The below diagram illustrates how the sequencer and verifiers fit together:
+
+![Propagation](./assets/propagation.svg)
+
+## Key Interactions In Depth
+
+### Deposits
+
+**Spec links:**
+
+- [Deposits](./deposits.md)
+
+Optimism supports two types of deposits: user deposits, and L1 attributes deposits. To perform a user deposit, users call the `depositTransaction` method on the `DepositFeed` contract. This in turn emits `TransactionDeposited` events, which the rollup node reads during block derivation.
+
+L1 attributes deposits are used to register L1 block attributes (number, timestamp, etc.) on L2 via a call to the L1 Attributes Predeploy. They cannot be initiated by users, and are instead added to L2 blocks automatically by the rollup node.
+
+Both deposit types are represented by a custom EIP-2718 transaction type on L2.
+
+### Block Derivation
+
+#### Overview
+
+The rollup chain can be deterministically derived given an L1 Ethereum chain. The fact that the entire rollup chain can be derived based on L1 blocks is _what makes OE a rollup_. This process can be represented as:
+
+```
+derive_rollup_chain(l1_blockchain) -> rollup_blockchain
+```
+
+Optimism's block derivation function is designed such that it:
+
+- Requires no state other than what is easily accessible using L1 and L2 execution engine APIs.
+- Supports sequencers and sequencer consensus.
+- Is resilient to sequencer censorship.
+
+#### Epochs and the Sequencing Window
+
+The rollup chain is subdivided into epochs. There is a 1:1 correspondence between L1 block numbers and epoch numbers. For L1 block number `n`, there is a corresponding rollup epoch `n` which can only be derived after L1 block number `n` is added to the L1 chain. 
+
+Each epoch contains at least one block. Every block in the epoch contains an L1 info transaction which contains contextual about L1 such as the block hash and timestamp. The first block in the epoch also contains all deposits initiated via the `DepositFeed` contract on L1. All subsequent blocks contain batches of transactions submitted directly to the sequencer.
+
+The sequencer can create blocks for future epochs, and reference them before confirmation on L1. This is what enables _fast transaction confirmations_ on the sequencer. To prevent censorship, however, the sequencer must batch-submit the blocks belonging to each epoch within a fixed number of L1 blocks. This submission window is called the _sequencing window_. The `epoch`s described above correspond to the first block in each sequencing window. Since transaction batches for a given epoch can be submitted anywhere within the sequencing window, verifiers must search all blocks within the window for transaction batches.
```

Comment:
A bit vague what "future epochs" means here.

The above is true if "future epoch"  == an epoch `n` such that the sequencing window `n-1` isn't finished yet.

It isn't true if "future epoch" == an epoch `n` such that the L1 block `n` has not been create yet.

In general a bit unsure about this paragraph. I think the idea you want to convey here is that to include transactions in a L2 block contained within a given epoch, the sequencer must submit a batch containing the transaction and the epoch number within the sequencing window of the epoch.

Reply:
This paragraph is supposed to describe how we get fast confirmations while also preventing censorship by the sequencer. "Future epochs" here means epochs for which the window isn't finished yet. How do you recommend I describe this distinction?

Reply:
I'll post my suggested rewrite for this section below, but explaining changes:

> All subsequent blocks contain batches of transactions submitted directly to the sequencer.

Bit confusing to use the word *batch* outside of batch submission.

> All subsequent blocks contain batches of transactions submitted directly to the sequencer.

This seems to imply the first block cannot contain sequenced transactions, which it can.

> To prevent censorship, however, the sequencer must batch-submit the
blocks belonging to each epoch within a fixed number of L1 blocks.

I think it's pretty confusing in what batch submission has to do with censorship prevention here.

Otherwise I think this part is tricky because batch submission isn't really detailed before.

---

```
The rollup chain is subdivided into epochs. There is a 1:1 correspondence between L1 block numbers and epoch numbers.
For L1 block number `n`, there is a corresponding rollup epoch `n` which can only be derived a "sequencing windows" has
passed, i.e. after L1 block number `n + SEQUENCING_WINDOW_SIZE` is added to the L1 chain.

Each epoch contains at least one block. Every block in the epoch contains an L1 info transaction which contains
contextual information about L1 such as the block hash and timestamp. The first block in the epoch also contains all
deposits initiated via the `DepositFeed` contract on L1. All L2 blocks can also contain *sequenced transactions*, i.e.
transactions submitted directly to the sequencer.

Whenever the sequencer creates a new L2 block for a given epoch, it must submit it to L1 as part of a *batch*, within
the epoch's sequencing window (i.e. the batch must land before L1 block `n + SEQUENCING_WINDOW_SIZE`). These batches are
(along with the `TransactionDeposited` L1 events) what allows the derivation of the L2 chain from the L1 chain.

The sequencer does not need for a L2 block to be batch-submitted to L1 in order to build on top of it. In fact, batches
typically contain multiple L2 blocks worth of sequenced transaction. This is what enables
_fast transaction confirmations_ on the sequencer.

Since transaction batches for a given epoch can be submitted anywhere within the sequencing window, verifiers must
search all blocks within the window for transaction batches. This protects against the uncertainty of transaction
inclusion of L1. This uncertainty is also why we need the sequencing window in the first place: otherwise the sequencer
could retroactively add blocks to an old epoch, and validators wouldn't know when they can finalize an epoch.

The sequencing window also prevents censorship by the sequencer: deposits made on a given L1 block will be included in
the L2 chain at worst after `SEQUENCING_WINDOW_SIZE` L1 blocks have passed.
```

Reply:
I applied your rewrite.